### ISSUE TITLE
inaccurate example on the :not() page

### ISSUE THREAD

**Issue by michal-amichai** [mdn/content#24277]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/CSS/:not

### What specific section or headline is this issue about?

Description

### What information was incorrect, unhelpful, or incomplete?

last bullet:
> If any selector passed to the `:not()` pseudo-class is invalid or not supported by the browser, the whole rule will be invalidated. The effective way to overcome this behavior is to use [`:is`](https://developer.mozilla.org/en-US/docs/Web/CSS/:is) pseudo-class, which accepts a forgiving selector list. For example `:not(.foo, :invalid-pseudo-class)` will invalidate a whole rule, but `:is(:not(.foo), :not(:invalid-pseudo-class))` will match any element that isn't `.foo`.

### What did you expect to see?

the example is inaccurate, as `:not(.foo, :invalid-pseudo-class)` matches NEITHER `.foo` NOR `:invalid-pseudo-class`
whereas `:is(:not(.foo), :not(:invalid-pseudo-class))` matches anything that simply isn't `.foo` (disregarding the second condition)
it will also match elements that weren't intended to initially, thus isn't a very good example of overcoming the issue.. :)

### Do you have any supporting links, references, or citations?

made a [pen example](https://codepen.io/michal-amichai/pen/oNMRyWG)

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/css/_colon_not`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/:not
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_not/index.md
* Last commit: https://github.com/mdn/content/commit/26a33b20edc0facf5a7b17ad0bf3190d847e366e
* Document last modified: 2022-12-15T20:42:42.000Z

</details>

**Comment by tlylt** [comment]
Hi @michal-amichai, I think the example could be confusing but ultimately an accurate one.

The example that you think is inaccurate is:
> For example :not(.foo, :invalid-pseudo-class) will invalidate a whole rule, but :is(:not(.foo), :not(:invalid-pseudo-class)) will match any element that isn't .foo.

You said that:
> as :not(.foo, :invalid-pseudo-class) matches NEITHER .foo NOR :invalid-pseudo-class

This is true, it does not match .foo and does not match :invalid-pseudo-class. In fact, it matches nothing because this rule is invalid. This is accurately stated:
> For example :not(.foo, :invalid-pseudo-class) will invalidate a whole rule

---

The second point is that
> whereas :is(:not(.foo), :not(:invalid-pseudo-class)) matches anything that simply isn't .foo (disregarding the second condition)

This is true, it only matches elements that do not have .foo, because :is() is forgiving. Which is accurately stated:
> but :is(:not(.foo), :not(:invalid-pseudo-class)) will match any element that isn't .foo.

---

As for your point that
> it will also match elements that weren't intended to initially, thus isn't a very good example of overcoming the issue.. :)

I suppose you are referring to the fact that html and body elements will be unintentionally selected? But that is already stated in point 4:
> :not(.foo) will match anything that isn't .foo, including `html` and `body`

I hope the above clarifies your doubts if any. I think the point itself is in fact accurate, but I agree that a more realistic example would be great. I have constructed a simple example to illustrate, you can find the source code [here](https://github.com/tlylt/web-examples/tree/main/cssnot), and the rendered page [here](https://tlylt.github.io/web-examples/cssnot/index.html). I will file a follow-up PR to add an example to the MDN page to make it clear as well.

Cheers!

**Comment by michal-amichai** [comment]
Hi @tlylt,
Thank you for your answer, and sorry for the late reply.
I think you missed my point, 
I didn't refer to html or body being selected,
the inaccuracy I'm talking about refers to the **initial intent**: 
`:not(A, B)` - means I want to select elements that are **neither** A **nor** B
`:is(:not(A),:not(B))` - selects **everything** that isn't A, **including B**
thus isn't overcoming the issue, it results in **a different selection than initially intended**.

an easy way to understand it:
```
<li class=red>
<li class=blue>
<li class=green>
```
`li:not(.red,.blue)` - selects green
`li:is(:not(.red),:not(.blue))` - selects everything :)
(you can see it live [here](https://codepen.io/michal-amichai/pen/yLxVJwV))

**Comment by tlylt** [comment]
Hi @michal-amichai, this is interesting. Thank you for making the point clearer, I indeed have misinterpreted the inaccuracy you are alluding to. While the part about how the whole rule will no longer be invalidated with `:is()`, the targeted elements will indeed be different if the second selector is in fact, valid.

I went on to think about whether there's a way to overcome the invalidation using `:is()` and here are my findings:

First of all, one detail of the `:not()` selector mentioned on the same page is this:
> You can negate several selectors at the same time. Example: :not(.foo, .bar) is equivalent to :not(.foo):not(.bar).

I did a test and the above is true (see demo [here](https://tlylt.github.io/web-examples/cssnot2/index.html)).
```html
    <div>
      <p>Hello World without any class</p>
      <p class="foo">Hello World with .foo</p>
      <p class="bar">Hello World with .bar</p>
      <p class="foo bar">Hello World with .foo and .bar</p>
    </div>
```
```css
p:not(.foo, .bar) {
  color: red;
  border-bottom: dotted thin currentcolor;
  font-style: italic;
}

p:not(.foo):not(.bar) {
  color: green;
  border-top: dotted thin currentcolor;
  font-weight: bold;
}
```
However, when they are used in `:is()`, like the following, the two will select different elements (see another demo [here](https://tlylt.github.io/web-examples/cssnot3/index.html))
```html
    <div>
      <p>Hello World without any class</p>
      <p class="foo">Hello World with .foo</p>
      <p class="bar">Hello World with .bar</p>
      <p class="foo bar">Hello World with .foo and .bar</p>
    </div>
```
```css
/* Select all paragraph elements that either do not have a foo or a bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
/* <p class="foo"> because it is not .bar */
/* <p class="bar"> because it is not .foo */
p:is(:not(.foo), :not(.bar)) {
  text-decoration: underline;
}

/* Select all paragraph elements that do not have a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
p:is(:not(.foo):not(.bar)) {
  color: green;
  border-top: dotted thin currentcolor;
}
```
Comparing the above, the conclusion that I can gather is that doing something like the could work to target the same elements while avoiding the whole rule being invalidated. (See the above demo [here](https://tlylt.github.io/web-examples/cssnot3/index.html))

```css
/* Select all paragraph elements that do not have a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
p:not(.foo, .bar) {
  color: red;
  border-bottom: dotted thin currentcolor;
  font-style: italic;
}

/* Select all paragraph elements that do not have a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
p:is(:not(.foo):not(.bar)) {
  color: green;
  border-top: dotted thin currentcolor;
}
```

To sum up, it seems to be that if I have `p:not(.foo, .bar)`, I can probably do `p:is(:not(.foo):not(.bar))` to prevent invalidation of the whole rule.

What do you think?

**Comment by michal-amichai** [comment]
indeed, `:not(.foo, .bar)` and `:not(.foo):not(.bar)` are equivalent,
so `:not(.foo, .bar)` and `:is(:not(.foo):not(.bar))` would be as well

**BUT**

it brings us back to where we started - 
`:not(.foo):not(.bar)` is a single selector, you're not using `:is`'s OR functionality here,
so if either `.foo` or `.bar` are invalid - the whole rule still fails..
you can see it live [here](https://codepen.io/michal-amichai/pen/BaOWNqd)

```
<li class="foo">foo</li>
<li class="bar">bar</li>
<li class="foo bar">foo and bar</li>
<li>some other element</li>
```
```
li:before {content: "[before content]"}
.foo { text-decoration: underline 2px dotted red }
.bar:before { text-decoration: overline 2px dotted blue }

/* 2nd is invalid selector - the whole rule fails */
li:not(.foo,.bar:before) {
  font-weight: bold
}

/* 2nd is invalid selector - the whole rule still fails */
li:is(:not(.foo):not(.bar:before)) {
  font-weight: bold
}
```

ultimately, 
`:not` is a AND selector
`:is` is a OR selector
so I don't think you can use `:is` to split `:not`'s AND parts apart, at least not in a simple straightforward manner..

**Comment by tlylt** [comment]
> **BUT**
> 
> it brings us back to where we started - `:not(.foo):not(.bar)` is a single selector, you're not using `:is`'s OR functionality here, so if either `.foo` or `.bar` are invalid - the whole rule still fails.. you can see it live [here](https://codepen.io/michal-amichai/pen/BaOWNqd)

Ah, you got me there. That's a bummer 😢. I will probably think about this for a while and raise a PR to make this clear on the MDN page to benefit anyone else who's going to face this same issue.

**Comment by tlylt** [comment]
Hey @michal-amichai happy to share some progress on this, feel free to let me know what you think:

- instead of fixing `:not(.foo, :invalid-pseudo-class)` with `:is(:not(.foo), :not(:invalid-pseudo-class))`, we could do: `:not(:is(.foo, :invalid-pseudo-class))`
- this way, the invalid selector will be ignore by `:is` and the whole rule still functions to select what's left of the original `:not` selector.

I have constructed two examples to illustrate the above.
[E.g. 1 output](https://tlylt.github.io/web-examples/cssnot4/index.html)
- this example shows that `p:not(.foo, .bar)` selects the same thing as `p:not(:is(.foo, .bar))`
```css
/* Select all paragraph elements that do not have a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
p:not(.foo, .bar) {
  color: red;
  border-bottom: dotted thin currentcolor;
  font-style: italic;
}

/* Select all paragraph elements that either has a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class=""> */
/* e.g. ALL of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
p:is(.foo, .bar) {
  text-decoration: underline;
}

/* Select all paragraph elements that do not have a foo or a bar or both foo and bar in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="bar"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
p:not(:is(.foo, .bar)) {
  color: green;
  border-top: dotted thin currentcolor;
}
```
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>CSS :not() example</title>
    <link rel="stylesheet" href="index.css" />
  </head>
  <body>
    <h1>CSS :not() additional example 3</h1>
    <div>
      <p>Hello World without any class</p>
      <p class="foo">Hello World with .foo</p>
      <p class="bar">Hello World with .bar</p>
      <p class="foo bar">Hello World with .foo and .bar</p>
    </div>
    <script></script>
  </body>
</html>
```

[E.g. 2 output](https://tlylt.github.io/web-examples/cssnot5/index.html)
- this example shows that `p:not(.foo, :invalid-selector)` selects the same thing as `p:not(:is(.foo, :invalid-selector))`

```css
/* Select nothing since the whole rule is invalidated */
p:not(.foo, :invalid-selector) {
  color: red;
  border-bottom: dotted thin currentcolor;
  font-style: italic;
}

/* Select all paragraph elements that do not have a foo in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
/* <p class="bar"> */
p:not(.foo) {
  color: yellow;
  border-left: dotted thick currentcolor;
  font-weight: bold;
}

/* Select all paragraph elements that do not have a foo in their class */
/* e.g. NONE of the following will be selected: */
/* <p class="foo"> */
/* <p class="foo bar"> */
/* e.g. ALL of the following will be selected: */
/* <p class=""> */
/* <p class="bar"> */
p:not(:is(.foo, :invalid-selector)) {
  color: green;
  border-top: dotted thin currentcolor;
}
```

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>CSS :not() example</title>
    <link rel="stylesheet" href="index.css" />
  </head>
  <body>
    <h1>CSS :not() additional example 4</h1>
    <div>
      <p>Hello World without any class</p>
      <p class="foo">Hello World with .foo</p>
      <p class="bar">Hello World with .bar</p>
      <p class="foo bar">Hello World with .foo and .bar</p>
    </div>
    <script></script>
  </body>
</html>
```

One thing I realized is that this problem is inherently not solvable...... as in there's no such thing as overcoming the invalid selection. Even if the use of :is can help preserve the partial :not selection, the selection may no longer be what was intended for (because what's intended for is invalid).

Nonetheless, if the above is logical & reasonable, I will put up an update to fix the example with this more appropriate way to "overcome" the invalidation.

**Comment by michal-amichai** [comment]
oh wow! that's a nice one! great job! 
and it actually DOES fix the problem completely.
I updated the [demo](https://codepen.io/michal-amichai/pen/BaOWNqd) if you wanna see it live

the problem was that `:not()` doesn't accept pseudo selectors
so pseudo selectors that are otherwise valid, become invalid inside `:not()`
for example: `.bar::after` is valid, but `:not(.bar::after)` is not

but `:is()` isn't limited in that way,
`:is(.bar::after)` is actually a valid selector

and since `:not(A,B)` is equivalent to `:not(:is(A,B))`, using your solution solves the problem. 

also, `:is()` is a forgiving selector, so even if one of the selectors inside really **were** to be invalid (not supported by the browser, for example), it would just ignore that one, as expected. 
the solution wasn't meant to add support for invalid selectors, just to allow an already **valid** ones to be used inside `:not()`.

I would say it again, really nice job! :D
Cheers!
